### PR TITLE
Zio timeout bench and opt

### PR DESCRIPTION
**Pull request by eyalfa**
adresses #9211 

this PR eliminates one of the fibers used by the original implementation (by delegation to race) and directly uses the scheduler instead combined with `ZIO.async` and a mechanism favoring the happy case attempting to avoid the async cost when the effects completes immediately.

benchmark results:
```
Benchmark                          (n)   Mode  Cnt        Score        Error  Units
TimeoutBenchmark.zioBaseline         0  thrpt   10  9820895.810 ±  73559.828  ops/s
TimeoutBenchmark.zioBaseline       100  thrpt   10   698217.347 ±  15910.613  ops/s
TimeoutBenchmark.zioBaseline     10000  thrpt   10     8442.558 ±    201.020  ops/s
TimeoutBenchmark.zioTimeoutAlt       0  thrpt   10  1370528.184 ± 114946.956  ops/s
TimeoutBenchmark.zioTimeoutAlt     100  thrpt   10   359349.674 ±   3779.519  ops/s
TimeoutBenchmark.zioTimeoutAlt   10000  thrpt   10     8051.479 ±     30.176  ops/s
TimeoutBenchmark.zioTimeoutOrig      0  thrpt   10    89375.888 ±   3473.065  ops/s
TimeoutBenchmark.zioTimeoutOrig    100  thrpt   10    76895.544 ±   2726.181  ops/s
TimeoutBenchmark.zioTimeoutOrig  10000  thrpt   10     6001.324 ±     67.938  ops/s
```

### CODE REVIEW

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala [hunk 1]
```diff
@@ -2752,7 +2760,11 @@ object ZIOSpec extends ZIOBaseSpec {
                       Some(ZIO.unit)
                     }
                     .flatMap { _ =>
-                      ZIO.never
+                      //ZIO.never
```
Comment (kyri-petrou) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala:2763]: Remove commented code

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala [hunk 2]
```diff
@@ -2726,18 +2726,26 @@ object ZIOSpec extends ZIOBaseSpec {
...
-                        }
+                    .ensuring {
+                      ZIO.scopeWith { /*scope*/
+                        _ =>
+                          ZIO.async[Any, Nothing, Unit] { /*k*/
+                            _ =>
+                              Unsafe.unsafe { implicit unsafe =>
+                                runtime.unsafe.fork {
+                                  //this makes sure to interrupt this fiber once the external scope closes (which happens AFTER the entire spec executed)
+                                  //scope.addFinalizer(ZIO.succeed(k(ZIO.interrupt))) *>
```
Comment (kyri-petrou) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala:2737]: Remove commented code

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala (hunk 1 above)
Comment (kyri-petrou) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala:2763]: As above

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala [hunk 3]
```diff
@@ -2752,7 +2760,11 @@ object ZIOSpec extends ZIOBaseSpec {
                       Some(ZIO.unit)
                     }
                     .flatMap { _ =>
-                      ZIO.never
+                      //ZIO.never
+                      zio.Promise
+                        .make[Nothing, Any]
+                        .withFinalizer(_.interrupt)
+                        .flatMap(_.await)
```
Comment (kyri-petrou) [inline core-tests/shared/src/test/scala/zio/ZIOSpec.scala:2767]: I'm not sure I understand this change. AFAICT this is the same as doing `ZIO.never`. What's the reason behind it?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 4]
```diff
@@ -5504,7 +5506,8 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
   }
 
   final class TimeoutTo[-R, +E, +A, +B](self: ZIO[R, E, A], b: () => B) {
-    def apply[B1 >: B](f: A => B1)(duration: => Duration)(implicit
+    //todo: kept here for reference, drop this once the new impl is approved (makre sure to drop the referencing benchmarks in zio.TimeoutBenchmark)
+    def applyOrig[B1 >: B](f: A => B1)(duration: => Duration)(implicit
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala:5510]: Reminder to remove this method

File: test/jvm-native/src/main/scala/zio/test/TestClockPlatformSpecific.scala [hunk 5]
```diff
@@ -32,7 +32,7 @@ trait TestClockPlatformSpecific { self: TestClock.Test =>
       new Scheduler {
         def schedule(runnable: Runnable, duration: Duration)(implicit unsafe: Unsafe): Scheduler.CancelToken = {
           val fiber =
-            runtime.unsafe.fork((sleep(duration) *> ZIO.succeed(runnable.run())))
+            runtime.unsafe.fork(((sleep(duration) *> ZIO.succeed(runnable.run())).interruptible))
```
Comment (kyri-petrou) [inline test/jvm-native/src/main/scala/zio/test/TestClockPlatformSpecific.scala:35]: Why is this change needed by the way? Was there something wrong in the previous implementation or is it needed for the new one?

File: test/js/src/main/scala/zio/test/TestClockPlatformSpecific.scala [hunk 6]
```diff
@@ -26,7 +26,7 @@ trait TestClockPlatformSpecific { self: TestClock.Test =>
       new Scheduler {
         def schedule(runnable: Runnable, duration: Duration)(implicit unsafe: Unsafe): Scheduler.CancelToken = {
           val fiber =
-            runtime.unsafe.fork(sleep(duration) *> ZIO.succeed(runnable.run()))
+            runtime.unsafe.fork((sleep(duration) *> ZIO.succeed(runnable.run())).interruptible)
```
Comment (kyri-petrou) [inline test/js/src/main/scala/zio/test/TestClockPlatformSpecific.scala:29]: As above

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 7]
```diff
@@ -5527,6 +5530,105 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
...
+                ex.map(f)
+            }
+        }
+      }
+  }
+  private object TimeoutTo {
+    sealed trait BypassState[+E, +A]
+    case object BypassPossible                                 extends BypassState[Nothing, Nothing]
+    case object BypassDenied                                   extends BypassState[Nothing, Nothing]
+    case class BypassPendingResult[+E, +A](ex: zio.Exit[E, A]) extends BypassState[E, A]
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala:5631]: Change this to `final`

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 8]
```diff
@@ -5527,6 +5530,105 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
...
+                    if (bypassState.compareAndSet(BypassPossible, BypassDenied)) //won the race
+                      cb(ZIO.left(b()).ensuring(fib.interrupt *> fib.inheritAll))
+                    else if (bypassState.getPlain eq BypassDenied) //supervisor won
+                      cb(ZIO.left(b()).ensuring(fib.interrupt *> fib.inheritAll))
+                  },
+                  duration
+                )(zio.Unsafe.unsafe)
+
+              if (
+                bypassState.get /*Plain*/ () eq BypassDenied
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala:5567]: AFAICT this branch will only ever be called in cases that `duration` is extremely small or zero? Can't we just check if the duration = 0 before scheduling the cancellation and avoid this check?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 9]
```diff
@@ -5527,6 +5530,105 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
...
+                    //in either case the state is BypassDenied and we're in race with the scheduler, and we know for sure this entire effect will be completed via cb.
+                    //since cb CAN be invoked multiple times, we simply delegate the race to cb
+                    // * notice that changing the scheduler to use bypass as well will require modifying this logic as state may be BypassPendingResult, violating the assumption behind this logic.
+                    cancellable.apply()
+                    cb {
+                      (fib.inheritAll.as(Right(ex)))
+                    }
+                  } //else: fiber won, parent now owns cb and can bypass it
+                }(zio.Unsafe.unsafe)
+                fib.start(self)
```
Comment (eyalfa) [inline core/shared/src/main/scala/zio/ZIO.scala:5586]: @kyri-petrou @ghostdogpr ,
I suspect this creates a time-window in which the parent fiber doesn't respond to interrupts, 
this may happen when the `self` effect is cpu intensive and does not make any async calls.

is this a real concern?

Comment (kyri-petrou) [reply core/shared/src/main/scala/zio/ZIO.scala:5586]: I was going through this PR earlier in the week and came to this point as well. Honestly, I don't really think adding timeouts to CPU-bound work makes sense, but I don't think we can't ignore the fact that it might indeed exist in the wild

**Review by kyri-petrou — COMMENTED** [review]
Still need to dig a bit deeper into the logic, leaving some comments in the meantime. Also, is this PR still needed or should it be closed in favour of #9266

**Comment by eyalfa** [comment]
> Still need to dig a bit deeper into the logic, leaving some comments in the meantime. Also, is this PR still needed or should it be closed in favour of #9266

I think I'm going to close this one and focus on #9266 which needs some more work to be done.
since #9266 builds on this branch (benchmark code, few fixes to the tests), I will try to answer some of your questions though

File: core-tests/shared/src/test/scala/zio/ZIOSpec.scala (hunk 3 above)
Comment (eyalfa) [reply core-tests/shared/src/test/scala/zio/ZIOSpec.scala:2767]: thing is that `ZIO.never` in an uninterruptible region, can't be interrupted.
furthermore, this uninterruptible region is forked as a daemon, effectively preventing JVM shutdown.

This change replaces it with a `Promise.await' indirectly bound to the outer scope (the entire suit's scope), this guarantees that once the suite compltes JVM is able to shut down.

I've observed this when running `ZIOSpec` in Intellij.

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 8 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/ZIO.scala:5567]: I think it is possible, I'm not sure how it compares to the original code that always executed the forked sleep. i/ think the scheduler detects this and 'fires' immediately anyway.

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 9 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/ZIO.scala:5586]: I actually used to have such a usecase where we executed 'untrusted' user code and wanted to make sure we can bail out in case to trouble (we did have some interrupt like mechanism to terminate the user code)

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 7 above)
Comment (eyalfa) [reply core/shared/src/main/scala/zio/ZIO.scala:5631]: it inherits a sealed trait, does it matter that it's not final?

File: test/js/src/main/scala/zio/test/TestClockPlatformSpecific.scala (hunk 6 above)
Comment (eyalfa) [reply test/js/src/main/scala/zio/test/TestClockPlatformSpecific.scala:29]: not sure what you mean in this comment, issue was when this code is ececuted inside an uniterruptible region the forked fiber can't be interrupted (interrupt awaits the fiber after signalling it) resulting with the canceller hanging when executed.
This causes the test scheduler to behave differently than the real scheduler, in a really bad way.

File: test/jvm-native/src/main/scala/zio/test/TestClockPlatformSpecific.scala (hunk 5 above)
Comment (eyalfa) [reply test/jvm-native/src/main/scala/zio/test/TestClockPlatformSpecific.scala:35]: see my comment above
